From 9dd6e3622d81e5e07cbeba4642afebc093caa526 Mon Sep 17 00:00:00 2001
From: aiaimimi0920 <153103332+aiaimimi0920@users.noreply.github.com>
Date: Mon, 7 Sep 2026 21:53:22 +0800
Subject: [PATCH] feat(rendering): expose unjittered projection

Preserve exact camera identity and stable depth mapping under TAA.

Godot-Patch-Topic: rendering.unjittered-projection
---
 .../scene_shader_forward_clustered.cpp             |  1 +
 .../scene_forward_clustered_inc.glsl               |  6 ++++++
 .../renderer_rd/shaders/scene_data_inc.glsl        |  2 ++
 .../storage_rd/render_scene_data_rd.cpp            | 14 ++++++++++++++
 .../renderer_rd/storage_rd/render_scene_data_rd.h  |  2 ++
 servers/rendering/shader_preprocessor.cpp          |  1 +
 servers/rendering/shader_types.cpp                 |  2 ++
 7 files changed, 28 insertions(+)

diff --git a/servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp b/servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp
index fc8febca3c16206870dc8379493615cbf8f919d1..9a349da8995790710855231bad46b5b4da07a197 100644
--- a/servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp
+++ b/servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp
@@ -712,6 +712,7 @@ void SceneShaderForwardClustered::init(const String p_defines) {
 		actions.renames["VIEW_MATRIX"] = "read_view_matrix";
 		actions.renames["INV_VIEW_MATRIX"] = "inv_view_matrix";
 		actions.renames["PROJECTION_MATRIX"] = "projection_matrix";
+		actions.renames["UNJITTERED_PROJECTION_MATRIX"] = "npr_unjittered_projection";
 		actions.renames["INV_PROJECTION_MATRIX"] = "inv_projection_matrix";
 		actions.renames["MODELVIEW_MATRIX"] = "modelview";
 		actions.renames["MODELVIEW_NORMAL_MATRIX"] = "modelview_normal";
diff --git a/servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl b/servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl
index ed568dc1191f179589b238e46ec188a6763f9432..af58c0aa7207064d99301e14154a1c0065568cf3 100644
--- a/servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl
+++ b/servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl
@@ -5,6 +5,12 @@
 #define MAX_VOXEL_GI_INSTANCES 8
 #define MAX_VIEWS 2
 
+#ifdef USE_MULTIVIEW
+#define npr_unjittered_projection scene_data.unjittered_projection_matrix_view[ViewIndex]
+#else
+#define npr_unjittered_projection scene_data.unjittered_projection_matrix
+#endif
+
 #extension GL_KHR_shader_subgroup_ballot : enable
 #extension GL_KHR_shader_subgroup_arithmetic : enable
 
diff --git a/servers/rendering/renderer_rd/shaders/scene_data_inc.glsl b/servers/rendering/renderer_rd/shaders/scene_data_inc.glsl
index 0561cfa7657a9c21e32f5ad291b7adde4382db77..b61839e08f610a71474181802e1424ba0cb9b9ab 100644
--- a/servers/rendering/renderer_rd/shaders/scene_data_inc.glsl
+++ b/servers/rendering/renderer_rd/shaders/scene_data_inc.glsl
@@ -15,6 +15,7 @@
 struct SceneData {
 	mat4 projection_matrix;
 	mat4 inv_projection_matrix;
+	mat4 unjittered_projection_matrix;
 	mat3x4 inv_view_matrix;
 	mat3x4 view_matrix;
 
@@ -25,6 +26,7 @@ struct SceneData {
 	// only used for multiview
 	mat4 projection_matrix_view[MAX_VIEWS];
 	mat4 inv_projection_matrix_view[MAX_VIEWS];
+	mat4 unjittered_projection_matrix_view[MAX_VIEWS];
 	vec4 eye_offset[MAX_VIEWS];
 
 	// Used for billboards to cast correct shadows.
diff --git a/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.cpp b/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.cpp
index 7120d536facc3684c6dd19957d9ed7496b190007..b2348b40db1c96d5cc106294ff35c182133e2e4b 100644
--- a/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.cpp
+++ b/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.cpp
@@ -80,6 +80,14 @@ void RenderSceneDataRD::update_ubo(RID p_uniform_buffer, RSE::ViewportDebugDraw
 	// just for easy access..
 	UBO &ubo = ubo_data.ubo;
 	UBO &prev_ubo = ubo_data.prev_ubo;
+	// Preserve the camera mapping before temporal jitter, without subtractive
+	// reconstruction from the jittered matrix (which loses exact float identity).
+	Projection unjittered_correction;
+	unjittered_correction.set_depth_correction(flip_y);
+	RendererRD::MaterialStorage::store_camera(unjittered_correction * cam_projection, ubo.unjittered_projection_matrix);
+	for (uint32_t v = 0; v < view_count; v++) {
+		RendererRD::MaterialStorage::store_camera(unjittered_correction * view_projection[v], ubo.unjittered_projection_matrix_view[v]);
+	}
 
 	Projection correction;
 	correction.set_depth_correction(flip_y);
@@ -265,6 +273,12 @@ void RenderSceneDataRD::update_ubo(RID p_uniform_buffer, RSE::ViewportDebugDraw
 	if (calculate_motion_vectors) {
 		// Q : Should we make a complete copy or should we define a separate UBO with just the components we need?
 		memcpy(&prev_ubo, &ubo, sizeof(UBO));
+		Projection prev_unjittered_correction;
+		prev_unjittered_correction.set_depth_correction(true);
+		RendererRD::MaterialStorage::store_camera(prev_unjittered_correction * prev_cam_projection, prev_ubo.unjittered_projection_matrix);
+		for (uint32_t v = 0; v < view_count; v++) {
+			RendererRD::MaterialStorage::store_camera(prev_unjittered_correction * prev_view_projection[v], prev_ubo.unjittered_projection_matrix_view[v]);
+		}
 
 		Projection prev_correction;
 		prev_correction.set_depth_correction(true);
diff --git a/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.h b/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.h
index 85a9a1e9beb2f9d80da4e6133e1c3945aa3154e8..8f8c999e8159d35643483c46ccba8bdab311815a 100644
--- a/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.h
+++ b/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.h
@@ -118,6 +118,7 @@ private:
 	struct UBO {
 		float projection_matrix[16];
 		float inv_projection_matrix[16];
+		float unjittered_projection_matrix[16];
 		float inv_view_matrix[12];
 		float view_matrix[12];
 
@@ -127,6 +128,7 @@ private:
 
 		float projection_matrix_view[RendererSceneRender::MAX_RENDER_VIEWS][16];
 		float inv_projection_matrix_view[RendererSceneRender::MAX_RENDER_VIEWS][16];
+		float unjittered_projection_matrix_view[RendererSceneRender::MAX_RENDER_VIEWS][16];
 		float eye_offset[RendererSceneRender::MAX_RENDER_VIEWS][4];
 
 		float main_cam_inv_view_matrix[16];
diff --git a/servers/rendering/shader_preprocessor.cpp b/servers/rendering/shader_preprocessor.cpp
index 178e379f3d670151bee332f21a8b4a76d7d5b25e..f50e5710b94bdac42ffbc1cd5a71183874278a86 100644
--- a/servers/rendering/shader_preprocessor.cpp
+++ b/servers/rendering/shader_preprocessor.cpp
@@ -1355,6 +1355,7 @@ Error ShaderPreprocessor::preprocess(const String &p_code, const String &p_filen
 		if (rendering_method == "forward_plus") {
 			insert_builtin_define("CURRENT_RENDERER", _MKSTR(2), pp_state);
 			insert_builtin_define("HAS_MATERIAL_POST_LIGHT", _MKSTR(1), pp_state);
+			insert_builtin_define("HAS_UNJITTERED_PROJECTION_MATRIX", _MKSTR(1), pp_state);
 		} else if (rendering_method == "mobile") {
 			insert_builtin_define("CURRENT_RENDERER", _MKSTR(1), pp_state);
 		} else { // gl_compatibility
diff --git a/servers/rendering/shader_types.cpp b/servers/rendering/shader_types.cpp
index d4070aa14ffb902238009274b1badb297054269c..54feab0718f9b2d7f165cad173478a06824a2b14 100644
--- a/servers/rendering/shader_types.cpp
+++ b/servers/rendering/shader_types.cpp
@@ -117,6 +117,7 @@ ShaderTypes::ShaderTypes() {
 	shader_modes[RSE::SHADER_SPATIAL].functions["vertex"].built_ins["VIEW_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
 	shader_modes[RSE::SHADER_SPATIAL].functions["vertex"].built_ins["INV_VIEW_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
 	shader_modes[RSE::SHADER_SPATIAL].functions["vertex"].built_ins["PROJECTION_MATRIX"] = ShaderLanguage::TYPE_MAT4;
+	shader_modes[RSE::SHADER_SPATIAL].functions["vertex"].built_ins["UNJITTERED_PROJECTION_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
 	shader_modes[RSE::SHADER_SPATIAL].functions["vertex"].built_ins["INV_PROJECTION_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
 	shader_modes[RSE::SHADER_SPATIAL].functions["vertex"].built_ins["MODELVIEW_MATRIX"] = ShaderLanguage::TYPE_MAT4;
 	shader_modes[RSE::SHADER_SPATIAL].functions["vertex"].built_ins["MODELVIEW_NORMAL_MATRIX"] = ShaderLanguage::TYPE_MAT3;
@@ -189,6 +190,7 @@ ShaderTypes::ShaderTypes() {
 	shader_modes[RSE::SHADER_SPATIAL].functions["fragment"].built_ins["VIEW_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
 	shader_modes[RSE::SHADER_SPATIAL].functions["fragment"].built_ins["INV_VIEW_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
 	shader_modes[RSE::SHADER_SPATIAL].functions["fragment"].built_ins["PROJECTION_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
+	shader_modes[RSE::SHADER_SPATIAL].functions["fragment"].built_ins["UNJITTERED_PROJECTION_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
 	shader_modes[RSE::SHADER_SPATIAL].functions["fragment"].built_ins["INV_PROJECTION_MATRIX"] = constt(ShaderLanguage::TYPE_MAT4);
 	shader_modes[RSE::SHADER_SPATIAL].functions["fragment"].built_ins["VIEWPORT_SIZE"] = constt(ShaderLanguage::TYPE_VEC2);
 	shader_modes[RSE::SHADER_SPATIAL].functions["fragment"].built_ins["FOG"] = ShaderLanguage::TYPE_VEC4; // TODO consider adding to light shader

base-commit: 5ec4857b340b6284a18b49b2eda462bd250f219a
prerequisite-patch-id: b38459b409a3326aed65d2433e1341638171f884
prerequisite-patch-id: bdb0fe8d6ba7fff1b900ae33d9bec552ea61724c
prerequisite-patch-id: 4dcdddc73066f768cb97c56444abacb21bda061c
prerequisite-patch-id: 1f2fbc55f35da3e943dcbaeeb3c9a6d798621bf3
prerequisite-patch-id: 4dd1cec6888d5160c85be2072b9e5dd05411d007
prerequisite-patch-id: 60110ae09a2d09796ad65e339ee0b5abc2ce541a
prerequisite-patch-id: bdb0025d4207a7cf0e1466ca96b2824d238b3caf
prerequisite-patch-id: 8c55a8201fec417bc2629b903ea2d630aedab1d8
prerequisite-patch-id: 9dfaaac9d4a47edf9d1eccce8fa66b69e674a911
prerequisite-patch-id: 17fa223ec74d20c7792e45826a4f0d8ecd7f85d7
prerequisite-patch-id: 0fb9ef780b3f10a7e82968e44defda508734fa91
prerequisite-patch-id: 2af824bf187ff13746bd620b1032f57d166c0114
prerequisite-patch-id: dc8aa901272b27086c9045fc0b9335986d3a8128
prerequisite-patch-id: 0ce30d5045f596dbc6bf72cbcd9bb16472513013
prerequisite-patch-id: 50d820c0fb3ac27ddef8f76b133637558f8bcc8e
prerequisite-patch-id: 69b2baab747aed9362c0ca4c161fc02e5970db44
prerequisite-patch-id: 41b4bd5af73a2a33c2e64048b72e0d4f3c664f19
prerequisite-patch-id: 67083f15ed3c32a690794c698729f213a5245e20
prerequisite-patch-id: 14ad80fb2ee81b00b9194186196bba9d350bc28e
prerequisite-patch-id: 7d791289954a2144845e9b843451c6c9717daf6a
prerequisite-patch-id: d5e37dc01427a0639bba6123e6afee995d62ef4a
prerequisite-patch-id: fa2f2bf9521e391400e0151a1a290fca10a3f124
prerequisite-patch-id: 2e5d11daca8780d00fe2d10677ac0c26098fae5c
prerequisite-patch-id: 06b6906be778d545881f82f2431b2204eb052a6c
prerequisite-patch-id: 6f98206b25188cf56bc55e3fbad4fff5da6d6d4e
prerequisite-patch-id: 9c598b9a701c8d9fb585255c239a2f64c998f3a0
prerequisite-patch-id: 148113e476c059293ce209839c317a4945cc1899
prerequisite-patch-id: 68f8ac60d0f60d3ac0bb1fe23aa79abd660c2009
prerequisite-patch-id: 470ac52175750bf3eadc5e0286a61411eb3f6ebf
prerequisite-patch-id: 075c9922d2a9fadb91e8c3402b4e7d2f7bf7f8a1
prerequisite-patch-id: 3b6bc3f484f1654f8bca99626ff2507a0e17a3ca
prerequisite-patch-id: d4fed2df8ba748ad28488271c1b2c42cc88f8c17
prerequisite-patch-id: 6a2c61a480ce572ae1c3692257cf60463d4d3cf9
